mountpoint-s3/src/data_cache/cache_directory.rs

```diff
@@ -0,0 +1,168 @@
+//! Provides functionality related to the inner cache directory Mountpoint creates or uses.
+//! Mountpoint attempts to cleanup the contents during mount and exit.
+//!
+//! Mountpoint uses a directory inside the user-provided cache directory
+//! to mitigate any impact from the user providing a directory that already contains data.
+//! Using a new sub-directory minimizes the interference with the existing directory structure,
+//! and limits the risk from deleting or overwriting data to files written within this sub-directory.
+
+use std::fs;
+use std::io;
+use std::path::{Path, PathBuf};
+
+use thiserror::Error;
+
+/// Cache directory that has been created and emptied, and will be emptied when dropped.
+#[derive(Debug)]
+pub struct ManagedCacheDir {
+    managed_path: PathBuf,
+    /// Used to prevent double cleanup with `close(self)` and [Drop]
+    closed: bool,
+}
+
+#[derive(Debug, Error)]
+pub enum ManagedCacheDirError {
+    #[error("creation of cache sub-directory failed due to IO error: {0}")]
+    CreationFailure(#[source] io::Error),
+    #[error("cleanup of cache sub-directory failed due to IO error: {0}")]
+    CleanupFailure(#[source] io::Error),
+}
+
+impl ManagedCacheDir {
+    /// Create a new directory inside the provided parent path, cleaning it of any contents if it already exists.
+    pub fn new_from_parent<P: AsRef<Path>>(parent_path: P) -> Result<Self, ManagedCacheDirError> {
+        let managed_path = parent_path.as_ref().join("mountpoint-cache");
+
+        if let Err(mkdir_err) = fs::create_dir(&managed_path) {
+            match mkdir_err.kind() {
+                io::ErrorKind::AlreadyExists => (),
+                _kind => return Err(ManagedCacheDirError::CreationFailure(mkdir_err)),
+            }
+        }
+
+        let managed_cache_dir = Self {
+            managed_path,
+            closed: false,
+        };
+        managed_cache_dir.clean()?;
```

Comment:
Yep, silly last minute change on my part. I'll put the behavior back to "empty directory contents" for now since I already have that in place.

Thankfully tests also caught this.